### PR TITLE
Warn about old flutter plugin registration format

### PR DESCRIPTION
**Pull request by jonasfj**
@amirh, does this look sane to you?

I'll write up test if you're happy with this, but is this the tests we want to do and is the wording of the messages reasonable to you?

### CODE REVIEW

File: lib/src/validator/flutter_plugin.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    // Report an error, if both the new and the old format is used.
+    if (usesOldPluginRegistrationFormat && usesNewPluginRegistrationFormat) {
+      errors.add('In pubspec.yaml the flutter.plugin.platforms plugin '
+          'registration format, cannot be used in combination with the old '
+          'flutter.plugin.{androidPackage,iosPrefix,pluginClass} '
+          'plugin registration format.');
+    }
+
+    // If the new plugin format is used, and the flutter SDK dependency allows
+    // SDKs older than 1.12, then this is going to be a problem.
```
Comment (amirh) [inline lib/src/validator/flutter_plugin.dart]: The minimal sdk should be 1.10 for the new schema I believe

File: lib/src/validator/flutter_plugin.dart [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      warnings.add('pubspec.yaml allows Flutter SDK version 1.11, which does '
+          'not support the flutter.plugin.platforms plugin registration '
+          'format. Please consider increasing the Flutter SDK requirement to '
+          '^1.12 (environment.sdk.flutter)');
+    }
+
+    // Warn if using the old plugin registration format.
+    if (usesOldPluginRegistrationFormat) {
+      warnings.add('The flutter.plugin.{androidPackage,iosPrefix,pluginClass} '
+          'plugin registration format is deprecated. Consider using the '
```
Comment (amirh) [inline lib/src/validator/flutter_plugin.dart]: nit: the term "plugin registration" may be a bit confusing here (I'd think it is talking about the generated plugin registrant), maybe saying: the {androidPackage,iosPrefix,pluginClass} keys are deprecated. Consider using the flutter.plugin.platforms key

File: lib/src/validator/flutter_plugin.dart [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          'flutter.plugin.platforms plugin registration format introduced in '
```
Comment (amirh) [inline lib/src/validator/flutter_plugin.dart]: We should probably add a go-link which have some explanation. TODO(amirh)

Comment (mit-mit) [reply lib/src/validator/flutter_plugin.dart]: Should probably be mentioned here? https://flutter.dev/docs/development/packages-and-plugins/developing-packages

**Comment by jonasfj** [comment]
@amirh, would you mind taking another look, I just:
 * Report error if using the `flutter.plugin.platforms` key along with the `flutter.plugin.{androidPackage,iosPrefix,pluginClass}` keys.
 * Report error if using the `flutter.plugin.platforms` key with Flutter SDK constraint that allows `1.9.x`. 
 * Added a few test cases.

Please confirm that:
 A) Flutter 1.9.x will be the last series to not support the new format, and,
 B) Flutter 1.10.0 will support the new platform format.
Is this correct? (It would be unfortunate for users if we hardcoded in the wrong constraints).

File: lib/src/validator/flutter_plugin.dart (hunk 3 above)
Comment (jonasfj) [reply lib/src/validator/flutter_plugin.dart]: @amirh, if you would like us to put a link into these warnings, please determine what that should be :)

File: lib/src/validator/flutter_plugin.dart [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import 'dart:async';
+
+import 'package:pub_semver/pub_semver.dart';
+
+import '../entrypoint.dart';
+import '../validator.dart';
+
+/// Validates that Flutter plugins doesn't use both new and old plugin format.
+///
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: Add a TODO to point to a document describing the formats when it becomes available.

File: lib/src/validator/flutter_plugin.dart [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+/// Warns if using the old plugin registration format.
+class FlutterPluginValidator extends Validator {
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: Perhaps: `FlutterPluginFormatValidator`

File: lib/src/validator/flutter_plugin.dart [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  FlutterPluginValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  Future validate() async {
+    final pubspec = entrypoint.root.pubspec;
+    // Ignore all packages that do not have the `flutter.plugin` property, or
+    // which does not have an SDK constraint on Flutter.
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: does -> do

File: lib/src/validator/flutter_plugin.dart [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    if (pubspec.fields['flutter'] is! Map ||
+        pubspec.fields['flutter']['plugin'] is! Map ||
+        pubspec.sdkConstraints['flutter'] == null) {
+      return;
+    }
+    final plugin = pubspec.fields['flutter']['plugin'] as Map;
+
+    // Determine if this uses the old format by checking if `flutter.plugin`
+    // contains any of the following keys.
+    final usesOldPluginFormat = {
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: Consider making the set const.

File: lib/src/validator/flutter_plugin.dart [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      errors.add('In pubspec.yaml the flutter.plugin.platforms key cannot be '
+          'used in combination with the old '
+          'flutter.plugin.{androidPackage,iosPrefix,pluginClass} keys.');
+    }
+
+    // If the new plugin format is used, and the flutter SDK dependency allows
+    // SDKs older than 1.10.0, then this is going to be a problem.
+    if (usesNewPluginFormat &&
+        pubspec.sdkConstraints['flutter'] != null &&
+        pubspec.sdkConstraints['flutter'].allows(Version.parse('1.9.999'))) {
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: I think you should test for empty intersection with a version constraint <1.10.0-0

File: lib/src/validator/flutter_plugin.dart [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    // Determine if this uses the new format by check if the:
+    // `flutter.plugin.platforms` keys is defined.
+    final usesNewPluginFormat = plugin['platforms'] != null;
+
+    // Report an error, if both the new and the old format is used.
+    if (usesOldPluginFormat && usesNewPluginFormat) {
+      errors.add('In pubspec.yaml the flutter.plugin.platforms key cannot be '
+          'used in combination with the old '
+          'flutter.plugin.{androidPackage,iosPrefix,pluginClass} keys.');
+    }
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: Consider early return

File: lib/src/validator/flutter_plugin.dart [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    if (usesNewPluginFormat &&
+        pubspec.sdkConstraints['flutter'] != null &&
+        pubspec.sdkConstraints['flutter'].allows(Version.parse('1.9.999'))) {
+      errors.add('pubspec.yaml allows Flutter SDK version 1.9.x, which does '
+          'not support the flutter.plugin.platforms key.\n'
+          'Please consider increasing the Flutter SDK requirement to '
+          '^1.10.0 (environment.sdk.flutter)');
+    }
+
+    // Warn if using the old plugin format.
```
Comment (sigurdm) [inline lib/src/validator/flutter_plugin.dart]: Comment seems redundant

File: test/validator/flutter_plugin_format_test.dart [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            },
+          },
+        },
+      };
+      await d.dir(appPath, [d.pubspec(pkg)]).create();
+      expectNoValidationError(flutterPluginFormat);
+    });
+  });
+
+  group('should consider a package invalid if it', () {
```
Comment (sigurdm) [inline test/validator/flutter_plugin_format_test.dart]: Nit: maybe split this group between warning and error cases.

**Pull request merged**